src/connector/src/parser/mod.rs

```diff
@@ -493,6 +495,147 @@ pub enum SpecificParserConfig {
     Bytes,
 }
 
+#[derive(Debug, Clone, Default)]
+pub struct ParserClientConfigList {
+    username: Option<String>,
+    password: Option<String>,
+}
+
+impl From<&HashMap<String, String>> for ParserClientConfigList {
+    fn from(props: &HashMap<String, String>) -> Self {
+        const SCHEMA_REGISTRY_USERNAME: &str = "schema.registry.username";
+        const SCHEMA_REGISTRY_PASSWORD: &str = "schema.registry.password";
+
+        ParserClientConfigList {
+            username: props.get(SCHEMA_REGISTRY_USERNAME).cloned(),
+            password: props.get(SCHEMA_REGISTRY_PASSWORD).cloned(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct AvroProperties {
+    pub use_schema_registry: bool,
+    pub row_schema_location: String,
+    pub upsert_primary_key: String,
+    pub client_config: ParserClientConfigList,
+    pub aws_auth_props: Option<AwsAuthProps>,
+    pub topic: String,
+    pub enable_upsert: bool,
+}
+
+#[derive(Default)]
+pub struct ProtobufProperties {
+    pub message_name: String,
+    pub use_schema_registry: bool,
+    pub row_schema_location: String,
+    pub aws_auth_props: Option<AwsAuthProps>,
+    pub client_config: ParserClientConfigList,
+    pub topic: String,
+}
+
+#[derive(Default)]
+pub struct CsvProperties {
+    pub delimiter: u8,
+    pub has_header: bool,
+}
+
+#[derive(Default)]
+pub enum EncodingProperties {
+    Avro(AvroProperties),
+    Protobuf(ProtobufProperties),
+    Csv(CsvProperties),
+    #[default]
+    None,
+}
+
+#[derive(Default)]
+pub enum ProtocolProperties {
+    Debezium,
+    Maxwell,
+    Canal,
+    #[default]
+    Plain,
+}
+
+pub struct ParserProperties {
+    pub encoding_config: EncodingProperties,
+    pub protocol_config: ProtocolProperties,
+}
+
+/// Requirements of each parser
+/// 1. CSV: delimiter, has header
+/// 2. AVRO/AVRO UPSERT: row schema location, use schema registry
+///    client info, topic, aws config, optional primary key
+/// 3. PROTOBUF: message name, row schema location, use schema registry
+///    client info, topic, aws config
+/// 4. DEBEZIUM AVRO: row schema location, topic, client info
+/// 5. Other: none
+
+impl ParserProperties {
+    pub fn new(
+        format: SourceFormat,
+        props: &HashMap<String, String>,
```

Comment:
can we access fields in `props` by `remove`? and report error if there is anything left when returning

Reply:
I think we can't. Currently we extract topic, `SchemaRegistryAuth`, `AwsAuthProps` from props and topic will be used later iiuc. For the auth part, it's hard to tell if any field is missing.